src/dataframe/dataframe.jl

```diff
@@ -1223,34 +1223,90 @@ Base.convert(::Type{DataFrame}, A::AbstractMatrix) = DataFrame(A)
+    end
+    if !(columns in (:identical, :equal, :intersect, :subset, :union))
+        throw(ArgumentError("`cols` keyword argument must be `:identical`, `:equal`, or `:intersect`"))
     end
     nrows, ncols = size(df)
+    targetrows = nrows + 1
     if ncols == 0 && row isa NamedTuple
         for (n, v) in pairs(row)
             setproperty!(df, n, fill!(Tables.allocatecolumn(typeof(v), 1), v))
         end
         return df
     end
-    # Only check for equal lengths, as an error will be thrown below if some names don't match
-    if columns === :equal && length(row) != ncols
-        # TODO: add tests for this case after the deprecation period
-        Base.depwarn("In the future push! will require that `row` has the same number" *
-                      " of elements as is the number of columns in `df`. " *
-                      "Use `columns=:intersect` to disable this check.", :push!)
+    if columns === :equal
+        # Only check for equal lengths if :equal is selected,
+        # as an error will be thrown below if some names don't match
+        if length(row) != ncols
+            # TODO: add tests for this case after the deprecation period
+            Base.depwarn("In the future push! will require that `row` has the same number" *
+                          " of elements as is the number of columns in `df`. " *
+                          "Use `columns=:intersect` to disable this check.", :push!)
+        end
+        if row isa NamedTuple
+            matching = all(x -> x[1]==x[2], zip(propertynames(row), _names(df)))
+            if !matching
+                Base.depwarn("columns=:equal as default is deprecated; in the future :identical " *
+                             "will be the default", :push!)
+            end
+        end
+    end
+    if columns === :identical
+        if length(row) != ncols
+            throw(ArgumentError("A pushed object must have the same number of elements" *
+                                " as a target data frame."))
+        end
+        if row isa NamedTuple
+            if any(x -> x[1] != x[2], zip(propertynames(row), _names(df)))
+                throw(ArgumentError("A pushed `NamedTuple` must have exactly the " *
+                                    "same column names and in the same order " *
+                                    "as a target data frame."))
+            end
+        end
+    end
+    if columns === :union
+        for k in keys(row)
+            if !hasproperty(df, k)
+                df[!, k] = Vector{Union{Missing, typeof(row[k])}}(missing, nrow(df))
+            end
+        end
     end
     current_col = 0
     try
         for (col, nm) in zip(_columns(df), _names(df))
             current_col += 1
-            push!(col, row[nm])
+            if columns === :union
+                val = get(row, nm, missing)
+                if val isa eltype(col)
+                    push!(col, val)
+                else
```

Comment:
This means that the column will be reallocated if the type is different, right? That's a big change. I suggest we only do that if we encounter a missing value and the column doesn't accept them. Otherwise, try assigning and let conversion fail if it needs to.

Reply:
What do you think about that approach: https://github.com/JuliaData/DataFrames.jl/pull/1958/files/755b96ee1ede1ce5bb2223e771f00100d50ddde8#r336946638
Ofc there we still can replace `similar` by `Tables.allocatecolumn`
Btw, what is the advantage of `Tables.allocatecolumn`? Does it also preserve array types like Pooled/Categorical arrays?

Reply:
> What do you think about that approach: https://github.com/JuliaData/DataFrames.jl/pull/1958/files/755b96ee1ede1ce5bb2223e771f00100d50ddde8#r336946638

Well currently `push!` doesn't reallocate columns when type doesn't match, so I think we should keep that in this PR. If we want to change it, better do it separately.

> Ofc there we still can replace `similar` by `Tables.allocatecolumn`
> Btw, what is the advantage of `Tables.allocatecolumn`? Does it also preserve array types like Pooled/Categorical arrays?

That's it.

Reply:
So I guess there's no difference to `similar(arr::Pooled, neweltype, newsize)`?

Reply:
Yes normally that should give the same result. But `Tables.allocatcolumn` works when you only have the element type and no existing column.

Reply:
@nalimilan - yes; this is a crucial decision for this PR. I am OK to change the way you propose, as I guess in practice in 99% of cases the `missing` will be a relevant widening that would be required by the user (but this means that we are diverging from `:union` in `vcat` and @rapus95 wanted us to stay consistent here where possible).

Reply:
how about some strict keyword which allows some granularity to decide whether the df that we push to is allowed to change the type of columns? Following options I'd recommend: :missing, :any, :none which by itself already are quite intuitive, :missing allowing missings, :any allowing to widen to any and :none allowing no changes at all. I would pick :missing as the default.
In my opinion having strict enabled is some sort of checking whether the input data has the right format. Though I'm not entirely sure whether that should be part of the push! API. I think it's preferable to convert types if necessary

Reply:
Why not. But better discuss that in another PR?

Reply:
I have decided to drop `:union` in this PR as it changed too much what currently `push!` did (i.e. it required mutating columns of the destination `df`). Let us leave this for a separate PR.